tests/test_web.py

```diff
@@ -225,16 +227,109 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+            }
+        )
+    ],
+    indirect=["environment"]
+)
+def test_sendmail_was_called_with_pdf_attachment(
+    environment, http, server, mock_render, email_invoice_request_body
+):
+    http.post("/invoice", json=email_invoice_request_body)
+    timestamp = datetime.fromisoformat(email_invoice_request_body['header']['timestamp'])
+    invoice_number = int(email_invoice_request_body['header']['number'])
+    invoice_name = f"{timestamp:%Y%m%d}-{invoice_number:04}-invoice.pdf"
+
+    email_content = server.sendmail.call_args.args[MESSAGE_ARGUMENT]
+    assert f"Content-Disposition: attachment; filename=\"{invoice_name}\"" in email_content
+
+
+@pytest.mark.parametrize(
+    "environment",
+    [
+        (
+            {
+                "INVOICE_UTILS_MAIL_SUBJECT": "test subject",
+                "INVOICE_UTILS_BODY_TEMPLATE_PATH": "test_template.html",
+                "INVOICE_UTILS_SENDER_EMAIL": "test@email.com",
+                "INVOICE_UTILS_INVOICE_DIR": DEFAULT_INVOICE_DIR
+            }
+        )
+    ],
+    indirect=["environment"]
+)
+def test_invoice_generated_within_specified_dir(
+    environment, http, server, mock_render, email_invoice_request_body
+):
+    http.post("/invoice", json=email_invoice_request_body)
+
+    assert environment.get("INVOICE_UTILS_INVOICE_DIR") in mock_render.call_args.args[1]
+
+
+@pytest.mark.parametrize(
+    "environment",
+    [
+        (
+            {
+                "INVOICE_UTILS_MAIL_SUBJECT": "test subject",
+                "INVOICE_UTILS_BODY_TEMPLATE_PATH": "test_template.html",
+                "INVOICE_UTILS_SENDER_EMAIL": "test@email.com",
+                "INVOICE_UTILS_INVOICE_DIR": "bad_directory"
+            }
+        )
+    ],
+    indirect=["environment"]
+)
+def test_invoice_generation_fails_if_directory_does_not_exist(
+    environment, http, server, mock_render, email_invoice_request_body
+):
+    res = http.post("/invoice", json=email_invoice_request_body)
+    assert res.status_code == 507
+    assert res.json() == {"detail": "Invoices directory not set up."}
+
+
+@pytest.mark.parametrize(
+    "environment",
+    [
+        (
+            {
+                "INVOICE_UTILS_MAIL_SUBJECT": "test subject",
+                "INVOICE_UTILS_BODY_TEMPLATE_PATH": "test_template.html",
+                "INVOICE_UTILS_SENDER_EMAIL": "test@email.com",
+                "INVOICE_UTILS_INVOICE_DIR": "bad_directory"
+            }
+        )
+    ],
+    indirect=["environment"]
+)
+def test_invoice_generation_fails_if_directory_does_not_exist(
+    environment, http, server, mocker, email_invoice_request_body
+):
+    with mocker.patch("invoice_utils.web.os.path.isdir", return_value=True):
+        with mocker.patch("invoice_utils.web.os.access", return_value=False):
+            res = http.post("/invoice", json=email_invoice_request_body)
+    assert res.status_code == 507
+    assert res.json() == {"detail": "Invoices directory does not have write access."}
```

Comment:
I like the test and the design. The error message could be less descriptive in this case. Should be something like 'insufficient rights to store invoice'.